File: wqr/wavelet_qqr.py
"""
Wavelet Quantile-on-Quantile Regression (WQQR) with P-values.

Combines MODWT wavelet decomposition with QQ Regression and
kernel-weighted local polynomial quantile regression.

Based on Adebayo, Özkan, Uzun Ozsahin, Eweade & Gyamfi (2025):
  Environmental Sciences Europe.
  doi:10.1186/s12302-025-01059-z
"""

import numpy as np
import pandas as pd
from dataclasses import dataclass
from typing import Optional

import statsmodels.api as sm
from statsmodels.regression.quantile_regression import QuantReg
from scipy import stats as sp_stats

from .utils import modwt_mra, gaussian_kernel, safe_print


@dataclass
class WQQRResult:
    """Container for Wavelet QQR results."""
    coef_matrix: np.ndarray     # (num × num) coefficient grid
    pval_matrix: np.ndarray     # (num × num) p-value grid
    qr_coef: np.ndarray         # (num,) standard QR slope coefficients
    qr_pval: np.ndarray         # (num,) standard QR p-values
    quantiles: np.ndarray
    band_name: str
    n_obs: int

    def summary(self):
        """Print summary."""
        n = len(self.quantiles)
        n_sig = (self.pval_matrix < 0.05).sum()
        lines = [
            "",
            "╔══════════════════════════════════════════════════════╗",
            "║  Wavelet QQR with P-values — Summary                ║",
            "╚══════════════════════════════════════════════════════╝",
            "",
            f"  Band        : {self.band_name}",
            f"  Observations: {self.n_obs}",
            f"  Quantiles   : {n}  [{self.quantiles[0]:.2f} … {self.quantiles[-1]:.2f}]",
            f"  Grid size   : {n} × {n} = {n*n} cells",
            f"  Coef range  : [{self.coef_matrix.min():.4f}, {self.coef_matrix.max():.4f}]",
            f"  Sig (5%)    : {n_sig} / {n*n}",
            "",
        ]
        safe_print("\n".join(lines))
        return self

    def avg_qqr_coef(self):
        """Average QQR coefficient at each y-quantile (compare with QR)."""
        return np.nanmean(self.coef_matrix, axis=0)


def _lprq(x, y, m, quantile_val, bandwidth=1.0):
    """
    Local polynomial quantile regression.

    For each grid point x_i, fits a weighted QR of y on (x - x_i)
    using Gaussian kernel weights, and extracts slope + p-value.
    """
    n = len(y)
    xx = np.linspace(x.min(), x.max(), m)
    slopes = np.zeros(m)
    pvals = np.zeros(m)

    for i in range(m):
        z = x - xx[i]
        w = sp_stats.norm.pdf(z / bandwidth)
        w = w / w.sum() * len(w) if w.sum() > 0 else w

        try:
            X_mat = sm.add_constant(z)
            model = QuantReg(y, X_mat)
            res = model.fit(q=quantile_val, max_iter=1000, p_tol=1e-5)
            slopes[i] = res.params[1]
            pvals[i] = res.pvalues[1] if res.pvalues is not None and len(res.pvalues) > 1 else np.nan
        except Exception:
            slopes[i] = np.nan
            pvals[i] = np.nan

    return slopes, pvals


def wavelet_qqr(y, x, quantile_step=0.05, wavelet="la8", J=5,
                band="long", bandwidth=1.0, verbose=True):
    """
    Wavelet Quantile-on-Quantile Regression with P-values.

    Parameters
    ----------
    y : array-like
        Dependent variable.
    x : array-like
        Independent variable.
    quantile_step : float
        Step size for quantile grid (default 0.05 → 19 quantiles).
    wavelet : str
        Wavelet family (R name).
    J : int
        Decomposition levels.
    band : str
        Which band: 'short', 'medium', 'long', 'all', or 'D1', 'D2', etc.
    bandwidth : float
        Gaussian kernel bandwidth for local QR.
    verbose : bool

    Returns
    -------
    WQQRResult
    """
    y = np.asarray(y, dtype=np.float64)
    x = np.asarray(x, dtype=np.float64)
    n = len(y)

    tau = np.arange(quantile_step, 1.0, quantile_step)
    num = len(tau)

    if verbose:
        print(f"Wavelet QQR  (wf={wavelet}, J={J}, band={band})")

    # Decompose
    det_y, _ = modwt_mra(y, wavelet=wavelet, level=J)
    det_x, _ = modwt_mra(x, wavelet=wavelet, level=J)

    # Select band
    band_lower = band.lower()
    if band_lower == "short":
        y_band = np.sum([det_y[i] for i in range(min(2, len(det_y)))], axis=0)
        x_band = np.sum([det_x[i] for i in range(min(2, len(det_x)))], axis=0)
    elif band_lower == "medium":
        idx = list(range(2, min(4, len(det_y))))
        if not idx:
            idx = [min(2, len(det_y) - 1)]
        y_band = np.sum([det_y[i] for i in idx], axis=0)
        x_band = np.sum([det_x[i] for i in idx], axis=0)
    elif band_lower == "long":
        idx = list(range(4, len(det_y)))
        if not idx:
            idx = [len(det_y) - 1]
        y_band = np.sum([det_y[i] for i in idx], axis=0)
        x_band = np.sum([det_x[i] for i in idx], axis=0)
    elif band_lower.startswith("d"):
        lvl = int(band_lower[1:]) - 1
        y_band = det_y[lvl]
        x_band = det_x[lvl]
    else:
        y_band = y
        x_band = x

    band_name = band.capitalize()
    yt = y_band[:n]
    xt = x_band[:n]

    # Standard QR
    if verbose:
        print("  Computing standard WQR …")
    X_qr = sm.add_constant(xt)
    qr_coef = np.zeros(num)
    qr_pval = np.zeros(num)
    for i, t in enumerate(tau):
        try:
            model = QuantReg(yt, X_qr)
            res = model.fit(q=t, max_iter=1000, p_tol=1e-5)
            qr_coef[i] = res.params[1]
            qr_pval[i] = res.pvalues[1] if res.pvalues is not None and len(res.pvalues) > 1 else np.nan
        except Exception:
            qr_coef[i] = np.nan
            qr_pval[i] = np.nan

    # QQR via local polynomial QR
    if verbose:
        print("  Computing WQQR grid …")
    coef_mat = np.zeros((num, num))
    pval_mat = np.zeros((num, num))

    for i, t in enumerate(tau):
        slopes, pvals = _lprq(xt, yt, num, t, bandwidth=bandwidth)
        coef_mat[:, i] = slopes
        pval_mat[:, i] = pvals
        if verbose and (i + 1) % max(1, num // 5) == 0:
            print(f"    Progress: {100*(i+1)//num}%")

    if verbose:
        print("  ✓ WQQR completed.")

    return WQQRResult(
        coef_matrix=coef_mat,
        pval_matrix=pval_mat,
        qr_coef=qr_coef,
        qr_pval=qr_pval,
        quantiles=tau,
        band_name=band_name,
        n_obs=n,
    )

File: R/wavelet_qqr.R
#' @title Wavelet Quantile-on-Quantile Regression with p-values (WQQR)
#'
#' @description Decomposes \eqn{y} and \eqn{x} with the MODWT, selects a
#'   frequency band, then for every (\eqn{\theta}, \eqn{\tau}) pair fits
#'   a kernel-weighted local linear quantile regression of \eqn{y_b} on
#'   \eqn{x_b}. Returns both the \code{coef_matrix} and the
#'   \code{pval_matrix} on the (\eqn{\theta}, \eqn{\tau}) grid, plus the
#'   standard band-level WQR slopes for comparison.
#'
#' @param y Numeric response.
#' @param x Numeric regressor.
#' @param quantile_step Numeric. Step of the \eqn{\tau}-grid.
#'   Default 0.05 -> 19 quantile levels.
#' @param wavelet Character. Wavelet filter.
#' @param J Integer. Decomposition depth.
#' @param band Character. One of \code{"short"}, \code{"medium"},
#'   \code{"long"}, \code{"all"} or a single detail-level label
#'   (\code{"D1"}, \code{"D2"}, ...).
#' @param bandwidth Numeric. Gaussian-kernel bandwidth (relative to the
#'   standardised band).
#' @param verbose Logical.
#'
#' @return Object of class \code{"wqqr_result"}.
#'
#' @references
#' Adebayo, T.S., Ozkan, O., Uzun Ozsahin, D., Eweade, B.S., Gyamfi, B.A.
#' (2025). Wavelet QQR.
#' \emph{Environmental Sciences Europe}.
#' \doi{10.1186/s12302-025-01059-z}.
#'
#' @examples
#' set.seed(1); n <- 128
#' x <- cumsum(rnorm(n)); y <- 0.4 * x + rnorm(n, sd = 0.5)
#' fit <- wavelet_qqr(y, x, quantile_step = 0.25,
#'                    wavelet = "la8", J = 4, band = "long",
#'                    verbose = FALSE)
#' print(fit)
#'
#' @export
#' @importFrom quantreg rq
#' @importFrom stats dnorm
wavelet_qqr <- function(y, x, quantile_step = 0.05,
                        wavelet = "la8", J = 5,
                        band = "long", bandwidth = 1.0,
                        verbose = TRUE) {
  y <- as.numeric(y); x <- as.numeric(x)
  n <- length(y)
  tau <- seq(quantile_step, 1 - quantile_step, by = quantile_step)
  num <- length(tau)
  if (verbose)
    message("WQQR (wf=", wavelet, ", J=", J, ", band=", band, ")")

  dec_y <- modwt_mra(y, wavelet = wavelet, level = J)
  dec_x <- modwt_mra(x, wavelet = wavelet, level = J)

  bl <- tolower(band)
  if (bl == "short") {
    idx <- seq_len(min(2L, J)); yb <- Reduce("+", dec_y$details[idx])
    xb <- Reduce("+", dec_x$details[idx])
  } else if (bl == "medium") {
    idx <- if (J >= 3) seq(3L, min(4L, J)) else min(2L, J)
    yb <- Reduce("+", dec_y$details[idx])
    xb <- Reduce("+", dec_x$details[idx])
  } else if (bl == "long") {
    idx <- if (J >= 5) seq(5L, J) else J
    yb <- Reduce("+", dec_y$details[idx])
    xb <- Reduce("+", dec_x$details[idx])
  } else if (substr(bl, 1, 1) == "d") {
    lvl <- as.integer(substr(bl, 2, nchar(bl)))
    yb <- dec_y$details[[lvl]]; xb <- dec_x$details[[lvl]]
  } else {
    yb <- y; xb <- x
  }
  band_name <- tools::toTitleCase(band)

  yb <- yb[seq_len(n)]; xb <- xb[seq_len(n)]

  # Standard WQR slope at each tau
  if (verbose) message("  WQR slopes ...")
  qr_coef <- numeric(num); qr_pval <- numeric(num)
  for (i in seq_len(num)) {
    fit <- tryCatch(quantreg::rq(yb ~ xb, tau = tau[i], method = "br"),
                    error = function(e) NULL)
    s <- if (!is.null(fit))
      tryCatch(summary(fit, se = "iid"), error = function(e) NULL)
      else NULL
    if (!is.null(s) && nrow(s$coefficients) >= 2) {
      qr_coef[i] <- s$coefficients[2, 1]
      qr_pval[i] <- s$coefficients[2, 4]
    } else {
      qr_coef[i] <- NA_real_; qr_pval[i] <- NA_real_
    }
  }

  # Local-linear QR grid
  if (verbose) message("  WQQR grid ...")
  xx <- seq(min(xb), max(xb), length.out = num)
  coef_mat <- pval_mat <- matrix(NA_real_, num, num,
                                 dimnames = list(round(tau, 4), round(tau, 4)))
  pct <- max(1L, num %/% 5L)
  for (i in seq_len(num)) {
    for (k in seq_len(num)) {
      z <- xb - xx[k]
      w <- stats::dnorm(z / bandwidth)
      sw <- sum(w); if (sw > 0) w <- w / sw * length(w)
      fit <- tryCatch(
        quantreg::rq(yb ~ z, tau = tau[i], weights = w, method = "br"),
        error = function(e) NULL)
      s <- if (!is.null(fit))
        tryCatch(summary(fit, se = "iid"), error = function(e) NULL)
        else NULL
      if (!is.null(s) && nrow(s$coefficients) >= 2) {
        coef_mat[k, i] <- s$coefficients[2, 1]
        pval_mat[k, i] <- s$coefficients[2, 4]
      }
    }
    if (verbose && i %% pct == 0)
      message("  ", round(100 * i / num), "%")
  }
  if (verbose) message("  done.")

  res <- list(coef_matrix = coef_mat, pval_matrix = pval_mat,
              qr_coef = qr_coef, qr_pval = qr_pval,
              quantiles = tau, band_name = band_name,
              n_obs = n, wavelet = wavelet, J = J,
              call = match.call(),
              method = "Wavelet Quantile-on-Quantile Regression (WQQR)")
  class(res) <- "wqqr_result"
  res
}


#' @export
print.wqqr_result <- function(x, ...) {
  cat("\nWavelet Quantile-on-Quantile Regression (WQQR)\n")
  cat(strrep("=", 48), "\n", sep = "")
  cat("  band =", x$band_name, "  wf =", x$wavelet,
      "  J =", x$J, "  N =", x$n_obs, "\n")
  cat("  q-grid =", length(x$quantiles), " cells =",
      length(x$quantiles)^2, "\n")
  n_sig <- sum(x$pval_matrix < 0.05, na.rm = TRUE)
  cat("  Coef range  : [", round(min(x$coef_matrix, na.rm = TRUE), 4),
      ",", round(max(x$coef_matrix, na.rm = TRUE), 4), "]\n")
  cat("  Sig at 5% : ", n_sig, "/", length(x$quantiles)^2, "\n", sep = "")
  invisible(x)
}

#' @export
summary.wqqr_result <- function(object, ...) { print(object); invisible(object) }

#' @export
plot.wqqr_result <- function(x, type = c("3d", "pvalue", "compare"),
                             colorscale = "Parula", ...) {
  type <- match.arg(type)
  fn <- switch(type, "3d" = plot_wqqr_surface,
               pvalue = plot_wqqr_pvalue_heatmap,
               compare = plot_wqr_vs_wqqr)
  fn(x, colorscale = colorscale, ...)
}
